### DISCUSSION TITLE
x-modelable on input

### DISCUSSION THREAD

**Discussion by Thanty** [alpinejs/alpine#3081] [7. Bugs (archived)]
Hi everyone!

I've ran into a problem with x-modelable directive.

I need to expose a property from input x-model to a higher-level component, but it doesn't work as expected. 
Right now you have to manually set property to make it work (as shown in docs https://alpinejs.dev/directives/modelable, hovewer, in case of input this seems weird).

Here's the example: https://codepen.io/Thanty/pen/XWEEqLd

**Comment by ercanertan** [comment]
Having the same issue..

**Reply by ekwoka** [reply]
Whats the issue?

The OPs example is just poorly designed. Why would you need to model it like that?

**Reply by ercanertan** [reply]
I have dynamic nested form component and top parent need to access specific child input value.
My workaround solution was using $ref instead x-modelable.
I hope the gif is clear to understand.
![latest](https://github.com/alpinejs/alpine/assets/13382249/fae700ee-8c4e-4017-a4b0-55fb3892f4a3)

**Reply by ekwoka** [reply]
So why does there need to be a middle point in the data there?

Just have the data only in the top one...

$ref definitely not the way to go for this. that seems way more convoluted.

It's a data tree.

Just use the data in the tree.

**Reply by ercanertan** [reply]
When saving the form, the fields should be in inner section of the component. The top text is just text but child is form input (the text only appear when hover). Some reason x-modelable didn't work for this case.

**Reply by ekwoka** [reply]
Yeah, so have the state on the outer thing.

This can be one component.

It doesn't need to be 5.

Maybe I can help you refactor it, but I just really don't get the issue you're having purely based on what's available right now.

**Reply by Thanty** [reply]
> The OPs example is just poorly designed.

In my case I have child component which have some state, that depends on input value.

I want to have an access to this state in a parent component, and **x-modelable** is great for that case.

But it doesn't work out of the box with input, so I have either to set state directly via **x-on:input** event, either duplicate **x-model** directive like in this example:

https://codepen.io/Thanty/pen/YzBjYeK

**Reply by ekwoka** [reply]
So, you have to use modelable correctly you mean? Modelable sets up 2 way data binding. Inputs have their own data to bind. So you don't put modelable on the input. You put it on the data. Because that's what you're modeling.

You could also just not have the intermediate value there.

**Reply by Thanty** [reply]
@ekwoka thanks for your replies! Is it the only correct way to use x-model twice?

```
<div id="parent-component" x-data="{text: ''}">
    <div id="child-component" x-data="{message: ''}" x-model="message" x-modelable="text">
      <input type="text" x-model="message">
    </div>
    <div>Text: <span x-text="text"></span></div>
</div>
```
First bind input text with `message` data, then set `message` data as a model and only after that bind `message` with `text`?

Also I'm not sure I understand the way you suggest not to use intermediate value. Inner component has and should have it's own data. And you cannot just duplicate data name, this won't work as expected.

```
<div id="parent-component" x-data="{message: ''}">
    <div id="child-component" x-data="{message: ''}">
      <input type="text" x-model="message">
    </div>
    <div>Text: <span x-text="message"></span></div>
</div>
```

You can skip inner data, but in that case there won't be inner component at all, just raw input.

```
<div id="parent-component" x-data="{message: ''}">
    <div id="child-component" class="not-a-component-just-element">
      <input type="text" x-model="message">
    </div>
    <div>Text: <span x-text="message"></span></div>
</div>
```
You cannot skip intermediate value, because inner component logic and data should remain unknown to the parent component.
Parent component should only know that inner component has some "public" value or "state" which can be received and bound to some variable.

**Comment by ercanertan** [comment]
@ekwoka 

Sure, I can access on the outer thing just like below before recursive loop starts.

``` @json( $filtered_fields[0]->inputValue ) ```

This looks to me ugly and not guarantee that first object will be label object.

I am working on multi level json object that can be convertible to php class during load. This label help to summarise current component without opening and checking what inside.

I accept that this is really rare case but I am here to point out that input value is not working with x-modellable.

**Reply by ekwoka** [reply]
I'm still not sure I understand what it being multi level or anything matters?

You can initialize the value from the place that it is, you can even give it's key.

Hell, you could just use a `@change` or `@input` event at the parent and check the formData

**Reply by ercanertan** [reply]
First things come to my mind was using $ref but using event could be the another way to do it..

**Reply by ercanertan** [reply]
I appreciate taking time to answer!
